devices/src/virtio/vsock/packet.rs

```diff
@@ -0,0 +1,344 @@
+/// present for data packets (VSOCK_OP_RW).
+///
+/// `VsockPacket` wraps these two buffers and provides direct access to the data stored
+/// in guest memory. This is done to avoid unnecessarily copying data from guest memory
+/// to temporary buffers, before passing it on to the vsock backend.
+///
+use byteorder::{ByteOrder, LittleEndian};
+
+use super::super::DescriptorChain;
+use super::defs;
+use super::{Result, VsockError};
+
+// The vsock packet header is defined by the C struct:
+//
+// ```C
+// struct virtio_vsock_hdr {
+//     le64 src_cid;
+//     le64 dst_cid;
+//     le32 src_port;
+//     le32 dst_port;
+//     le32 len;
+//     le16 type;
+//     le16 op;
+//     le32 flags;
+//     le32 buf_alloc;
+//     le32 fwd_cnt;
+// };
+// ```
+//
+// This structed will occupy the buffer pointed to by the head descriptor. We'll be accessing it
+// as a byte slice. To that end, we define below the offsets for each field struct, as well as the
+// packed struct size, as a bunch of `usize` consts.
+// Note that these offsets are only used privately by the `VsockPacket` struct, the public inteface
+// consisting of getter and setter methods, for each struct field, that will also handle the correct
+// endianess.
+
+// The vsock packet header struct size (when packed).
+const VSOCK_PKT_HDR_SIZE: usize = 44;
+
+// Source CID.
+const HDROFF_SRC_CID: usize = 0;
+
+// Destination CID.
+const HDROFF_DST_CID: usize = 8;
+
+// Source port.
+const HDROFF_SRC_PORT: usize = 16;
+
+// Destination port.
+const HDROFF_DST_PORT: usize = 20;
+
+// Data length (in bytes) - may be 0, if there is now data buffer.
+const HDROFF_LEN: usize = 24;
+
+// Socket type. Currently, only connection-oriented streams are defined by the vsock protocol.
+const HDROFF_TYPE: usize = 28;
+
+// Operation ID - one of the VSOCK_OP_* values; e.g.
+// - VSOCK_OP_RW: a data packet;
+// - VSOCK_OP_REQUEST: connection request;
+// - VSOCK_OP_RST: forcefull connection termination;
+// etc (see `super::defs::uapi` for the full list).
+const HDROFF_OP: usize = 30;
+
+// Additional options (flags) associated with the current operation (`op`).
+// Currently, only used with shutdown requests (VSOCK_OP_SHUTDOWN).
+const HDROFF_FLAGS: usize = 32;
+
+// Size (in bytes) of the packet sender receive buffer (for the connection to which this packet
+// belongs).
+const HDROFF_BUF_ALLOC: usize = 36;
+
+// Number of bytes the sender has received and consumed (for the connection to which this packet
+// belongs). For instance, for our Unix backend, this counter would be the total number of bytes
+// we have successfully written to a backing Unix socket.
+const HDROFF_FWD_CNT: usize = 40;
+
+/// The vsock packet, implemented as a wrapper over a virtq descriptor chain:
+/// - the chain head, holding the packet header; and
+/// - (an optional) data/buffer descriptor, only present for data packets (VSOCK_OP_RW).
+///
+pub struct VsockPacket {
+    hdr: *mut u8,
+    buf: Option<*mut u8>,
```

Comment:
I have some general concerns with these pointers to memory: the backing memory could definitely go away at some point before the VsockPacket disappears. VsockPacket should not exceed the lifetime of the backing memory or all the operations done to it should be otherwise unsafe. 

There is a comment around the unsafe code, but it's not entirely accurate.
It's true that you're using it in a way that's safe, as far as I can tell, but in the general case, this breaks memory safety rules, as the compiler can't tell if `hdr` and `buf` still point to valid memory.

```
// This is safe since bound checks have already been performed when creating the packet
// from the virtq descriptor.
```

I think adding a lifetime parameter to this type would improve things.

Reply:
Are you referring to the backing memory going away once the virtio buffers are marked used, and thusly their ownership returned to the driver? If that's so, I think a rigorously correct solution would have us synchronize the lifetime of virtio buffer wrappers (i.e. `DescriptorChain`) and `VsockPacket` in a non-trivial manner. That would require a considerable amount of work on our virtio handling code.

If you were referring to something else, could you please dive into a bit more detail? Thanks!

Reply:
No, that's exactly right. The lifetime of both `hdr` and `buf` depends on the logical lifetime of associated virtqueue descriptors. I haven't looked into an exact solution, but there should be some lifetime dependency between the VsockPacket and the memory it has pointers to. Not sure how easy that would be considering you hold direct pointers to memory instead of references.
Maybe something like this would work?
```
pub struct VsockPacket<'a, 'b> {
    hdr:  UnsafeCell<u8>,
    buf: Option<UnsafeCell<u8>>,
    phantom_hdr: PhantomData<&'a u8>,
    phantom_buf: PhantomData<&'b u8>,
}
```

Reply:
Well, then I don't think rust lifetimes are going to help, since descriptor lifetimes are not known at compilation time. I.e. the data pointed to by a descriptor lives up until the moment we add that descriptor to the used vring.

If I understand you correctly, you are referring to the rust lifetimes explicitly set for `DescriptorChain` (please correct me if I misunderstood). Those are actually not what they might appear to be, though. I.e. they aren't connected to the data lifetime (since, as I was saying, that can only be known at runtime), but are just there so `DescriptorChain` can hold a reference to `GuestMemory`. That pattern itself needs to be cleaned up, since references to `GuestMemory` make little sense (`GuestMemory` is just an `Arc`, so a reference to it is basically a statically-tracked reference to a dynamically-tracked reference).

Reply:
The descriptors themselves no, but the wrapping structure does have a lifetime.
```
let used_len = match VsockPacket::from_rx_virtq_head(&head)
```
This creates a new `VsockPacket` from `&head` by extracting the internal memory region of `head` and the next descriptor. However, after this creation succeeds, `VsockPacket` is completely disconnected from `head`, instead of owning a reference to it with a proper lifetime to it. So, you could add it to the used queue and the memory gets reclaimed by the guest, while the `VsockPacket` is still alive and usable from the VMM.

Speaking of which:
```
self.rxvq.add_used(&self.mem, head.index, used_len);
```
This needs to consume `head` and expire its lifetime instead of using `head.index`.

I think I'll turn this into a larger issue instead of commenting on this particular PR.